File: sdk/python/agenthub/agent.py
"""Core Agent class for the Agent Garden Full Code Python SDK.

Provides a builder-pattern API for defining agents programmatically.
The resulting agent config can be serialized to valid agent.yaml,
loaded from YAML, and deployed via the standard pipeline.
"""

from __future__ import annotations

import logging
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .deploy import DeployConfig, PromptConfig
from .memory import MemoryConfig
from .model import ModelConfig
from .tool import Tool, ToolConfig

logger = logging.getLogger(__name__)


@dataclass
class AgentConfig:
    """Configuration for an Agent Garden agent."""

    name: str
    version: str = "1.0.0"
    description: str = ""
    team: str = "default"
    owner: str = ""
    framework: str = "custom"
    model: ModelConfig | None = None
    tools: list[ToolConfig] = field(default_factory=list)
    prompts: PromptConfig | None = None
    memory: MemoryConfig | None = None
    guardrails: list[str] = field(default_factory=list)
    deploy: DeployConfig | None = None
    tags: list[str] = field(default_factory=list)
    knowledge_bases: list[str] = field(default_factory=list)


class Agent:
    """Full Code SDK entry point for defining agents programmatically.

    Supports a fluent builder pattern::

        agent = (
            Agent("my-agent", version="1.0.0", team="eng")
            .with_model(primary="claude-sonnet-4")
            .with_prompt(system="You are helpful.")
            .with_deploy(cloud="aws")
        )
    """

    def __init__(self, name: str, **kwargs: Any) -> None:
        self.config = AgentConfig(name=name, **kwargs)
        self._tools: list[Tool] = []
        self._middleware: list[Callable[..., Any]] = []
        self._hooks: dict[str, list[Callable[..., Any]]] = {}
        self._state: dict[str, Any] = {}

    # -----------------------------------------------------------------
    # Builder pattern methods (all return self for chaining)
    # -----------------------------------------------------------------

    def with_model(
        self,
        primary: str,
        fallback: str | None = None,
        **kwargs: Any,
    ) -> Agent:
        """Configure the LLM model."""
        self.config.model = ModelConfig(primary=primary, fallback=fallback, **kwargs)
        return self

    def with_tool(self, tool: Tool) -> Agent:
        """Add a tool to this agent."""
        self._tools.append(tool)
        self.config.tools.append(tool.to_config())
        return self

    def with_prompt(self, system: str, **kwargs: Any) -> Agent:
        """Set the system prompt (inline text or registry ref)."""
        self.config.prompts = PromptConfig(system=system, **kwargs)
        return self

    def with_memory(self, backend: str = "in_memory", **kwargs: Any) -> Agent:
        """Configure conversation memory."""
        self.config.memory = MemoryConfig(backend=backend, **kwargs)
        return self

    def with_guardrail(self, name: str) -> Agent:
        """Add a guardrail by name."""
        self.config.guardrails.append(name)
        return self

    def with_deploy(self, cloud: str = "local", **kwargs: Any) -> Agent:
        """Configure deployment target."""
        self.config.deploy = DeployConfig(cloud=cloud, **kwargs)
        return self

    def tag(self, *tags: str) -> Agent:
        """Add discovery tags."""
        self.config.tags.extend(tags)
        return self

    # -----------------------------------------------------------------
    # Advanced features
    # -----------------------------------------------------------------

    def use(self, middleware: Callable[..., Any]) -> Agent:
        """Add middleware (pre/post processing on every turn)."""
        self._middleware.append(middleware)
        return self

    def on(self, event: str, handler: Callable[..., Any]) -> Agent:
        """Register event hook (tool_call, turn_start, turn_end, error)."""
        self._hooks.setdefault(event, []).append(handler)
        return self

    @property
    def state(self) -> dict[str, Any]:
        """Typed state persisted across turns."""
        return self._state

    def route(self, message: str, context: dict[str, Any]) -> str | None:
        """Override for custom routing logic. Return tool/agent name or None."""
        return None

    def select_tools(self, message: str) -> list[Tool]:
        """Override for dynamic tool selection based on input."""
        return list(self._tools)

    # -----------------------------------------------------------------
    # Serialization
    # -----------------------------------------------------------------

    def to_yaml(self) -> str:
        """Serialize to valid agent.yaml content."""
        from .yaml_utils import agent_to_yaml

        return agent_to_yaml(self)

    @classmethod
    def from_yaml(cls, yaml_content: str) -> Agent:
        """Load agent from YAML string."""
        from .yaml_utils import yaml_to_agent

        return yaml_to_agent(yaml_content)

    @classmethod
    def from_file(cls, path: str) -> Agent:
        """Load agent from YAML file path."""
        content = Path(path).read_text(encoding="utf-8")
        return cls.from_yaml(content)

    def save(self, path: str) -> None:
        """Save agent.yaml to disk."""
        Path(path).write_text(self.to_yaml(), encoding="utf-8")
        logger.info("Saved agent config to %s", path)

    # -----------------------------------------------------------------
    # Deployment
    # -----------------------------------------------------------------

    def deploy(self, target: str = "local") -> dict[str, Any]:
        """Deploy this agent (wraps garden deploy).

        Returns deploy info dict. Currently a placeholder that saves the
        YAML and returns metadata; the real implementation delegates to
        the engine.DeployEngine pipeline.
        """
        logger.info("Deploying agent '%s' to target '%s'", self.config.name, target)
        return {
            "agent": self.config.name,
            "version": self.config.version,
            "target": target,
            "status": "pending",
        }

    def validate(self) -> list[str]:
        """Validate the agent config. Returns list of error messages (empty = valid)."""
        from .validation import validate_agent

        return validate_agent(self)

    def __repr__(self) -> str:
        return (
            f"Agent(name={self.config.name!r}, version={self.config.version!r}, "
            f"framework={self.config.framework!r})"
        )

File: sdk/typescript/src/agent.ts
/** Agent builder — TypeScript equivalent of the Python SDK Agent class. */

import { Model } from "./model";
import { Tool } from "./tool";
import { Memory } from "./memory";
import type { MemoryConfig } from "./memory";
import type {
  AgentConfig,
  CloudType,
  DeployConfig,
  FrameworkType,
  McpServerRef,
  PromptConfig,
  SubagentRef,
  ToolConfig,
} from "./types";
import { agentToYaml } from "./yaml";

export interface AgentOptions {
  version?: string;
  description?: string;
  team?: string;
  owner?: string;
  framework?: FrameworkType;
  tags?: string[];
}

export class Agent {
  private _name: string;
  private _version: string;
  private _description: string;
  private _team: string;
  private _owner: string;
  private _framework: FrameworkType;
  private _tags: string[];
  private _model: Model | null = null;
  private _tools: ToolConfig[] = [];
  private _subagents: SubagentRef[] = [];
  private _mcpServers: McpServerRef[] = [];
  private _prompts: PromptConfig = {};
  private _guardrails: string[] = [];
  private _deploy: DeployConfig | null = null;
  private _knowledgeBases: Array<{ ref: string }> = [];
  private _memory: Memory | null = null;
  private _middlewares: Array<(msg: string, ctx: Record<string, unknown>) => Record<string, unknown>> = [];
  private _handlers: Record<string, Array<(...args: unknown[]) => void>> = {};
  private _state: Record<string, unknown> = {};

  constructor(name: string, opts: AgentOptions = {}) {
    this._name = name;
    this._version = opts.version ?? "1.0.0";
    this._description = opts.description ?? "";
    this._team = opts.team ?? "default";
    this._owner = opts.owner ?? "";
    this._framework = opts.framework ?? "custom";
    this._tags = opts.tags ?? [];
  }

  withModel(primary: string, opts?: { fallback?: string; temperature?: number; maxTokens?: number }): this {
    const model = new Model(primary);
    if (opts?.fallback) model.fallback(opts.fallback);
    if (opts?.temperature !== undefined) model.temperature(opts.temperature);
    if (opts?.maxTokens !== undefined) model.maxTokens(opts.maxTokens);
    this._model = model;
    return this;
  }

  withTool(tool: Tool): this {
    this._tools.push(tool.toConfig());
    return this;
  }

  withSubagent(ref: string, opts?: { name?: string; description?: string }): this {
    this._subagents.push({ ref, ...opts });
    return this;
  }

  withMcpServer(ref: string, transport?: "stdio" | "sse" | "streamable_http"): this {
    this._mcpServers.push({ ref, transport });
    return this;
  }

  withPrompt(system: string): this {
    this._prompts = { system };
    return this;
  }

  withGuardrail(name: string): this {
    this._guardrails.push(name);
    return this;
  }

  withDeploy(cloud: CloudType, opts?: Partial<Omit<DeployConfig, "cloud">>): this {
    this._deploy = { cloud, ...opts };
    return this;
  }

  withMemory(backend: string, opts?: Partial<MemoryConfig>): this {
    this._memory = Memory.fromConfig({ backend, ...opts });
    return this;
  }

  tag(...tags: string[]): this {
    this._tags.push(...tags);
    return this;
  }

  use(middleware: (msg: string, ctx: Record<string, unknown>) => Record<string, unknown>): this {
    this._middlewares.push(middleware);
    return this;
  }

  on(
    event: "tool_call" | "turn_start" | "turn_end" | "error",
    handler: (...args: unknown[]) => void
  ): this {
    if (!this._handlers[event]) this._handlers[event] = [];
    this._handlers[event].push(handler);
    return this;
  }

  get state(): Record<string, unknown> {
    return { ...this._state };
  }

  route(_message: string, _context: Record<string, unknown>): string | null {
    return null;
  }

  selectTools(_message: string): Tool[] {
    return [];
  }

  toConfig(): AgentConfig {
    if (!this._model) throw new Error("Model is required — call .withModel()");
    if (!this._deploy) throw new Error("Deploy config is required — call .withDeploy()");

    const config: AgentConfig = {
      name: this._name,
      version: this._version,
      description: this._description,
      team: this._team,
      owner: this._owner,
      framework: this._framework,
      tags: this._tags,
      model: this._model.toConfig(),
      tools: this._tools,
      subagents: this._subagents.length > 0 ? this._subagents : undefined,
      mcp_servers: this._mcpServers.length > 0 ? this._mcpServers : undefined,
      knowledge_bases: this._knowledgeBases.length > 0 ? this._knowledgeBases : undefined,
      prompts: this._prompts,
      guardrails: this._guardrails,
      deploy: this._deploy,
    };

    if (this._memory) {
      config.memory = this._memory.toConfig();
    }

    return config;
  }

  toYaml(): string {
    return agentToYaml(this.toConfig());
  }

  validate(): string[] {
    const errors: string[] = [];
    if (!this._name) errors.push("name is required");
    if (!this._model) errors.push("model is required — call .withModel()");
    if (!this._deploy) errors.push("deploy config is required — call .withDeploy()");
    if (!this._team) errors.push("team is required");
    return errors;
  }

  static fromYaml(yaml: string): Agent {
    // Parse minimal YAML by extracting key: value pairs
    // For production use, a proper YAML parser should be used
    const lines = yaml.split("\n");
    const getValue = (key: string): string | undefined => {
      const line = lines.find((l) => l.trimStart().startsWith(`${key}:`));
      if (!line) return undefined;
      const val = line.slice(line.indexOf(":") + 1).trim();
      return val.replace(/^["']|["']$/g, "");
    };

    const name = getValue("name") ?? "unnamed-agent";
    const version = getValue("version");
    const description = getValue("description");
    const team = getValue("team");
    const owner = getValue("owner");
    const framework = getValue("framework") as FrameworkType | undefined;

    const agent = new Agent(name, {
      version: version ?? "1.0.0",
      description,
      team: team ?? "default",
      owner: owner ?? "",
      framework: framework ?? "custom",
    });

    // Parse model section
    const primary = getValue("primary");
    const fallback = getValue("fallback");
    if (primary) {
      agent.withModel(primary, fallback ? { fallback } : undefined);
    }

    // Parse deploy section
    const cloud = getValue("cloud") as CloudType | undefined;
    const runtime = getValue("runtime");
    if (cloud) {
      agent.withDeploy(cloud, runtime ? { runtime } : undefined);
    }

    return agent;
  }

  static async fromFile(path: string): Promise<Agent> {
    const { readFile } = await import("fs/promises");
    const content = await readFile(path, "utf-8");
    return Agent.fromYaml(content);
  }

  async save(path: string): Promise<void> {
    const { writeFile } = await import("fs/promises");
    await writeFile(path, this.toYaml(), "utf-8");
  }

  async deploy(target?: string): Promise<import("./deploy").DeployResult> {
    const { deploy: deployFn } = await import("./deploy");
    return deployFn(this.toConfig(), target);
  }
}
